File: services/category.go
package services

import (
	"context"

	"github.com/mohan9182/financer/auth"
	"github.com/mohan9182/financer/api"
	"github.com/mohan9182/financer/repository"
)

type CategoryService struct {
	repo *repository.CategoryRepository
}

func NewCategoryService(repo *repository.CategoryRepository) *CategoryService {
	return &CategoryService{repo: repo}
}

func (s *CategoryService) ListCategories(ctx context.Context, msg *api.ListCategoriesRequest) (*api.ListCategoriesResponse, error) {
	userID := ctx.Value(auth.UserIDKey).(string)
	result, err := s.repo.List(ctx, userID, msg.PageSize, msg.PageToken)
	if err != nil {
		return nil, ServerError("%v", err)
	}

	return &api.ListCategoriesResponse{
		Categories:    result.Categories,
		NextPageToken: result.NextPageToken,
	}, nil
}

func (s *CategoryService) CreateCategories(ctx context.Context, msg *api.CreateCategoriesRequest) (*api.BulkOperationResponse, error) {
	userID := ctx.Value(auth.UserIDKey).(string)

	var inputs []repository.CreateCategoryInput
	for _, c := range msg.Categories {
		if c.Name == "" {
			continue
		}
		inputs = append(inputs, repository.CreateCategoryInput{
			Name:   c.Name,
			UserID: userID,
		})
	}

	_, errs := s.repo.Create(ctx, inputs)
	if len(errs) > 0 {
		return &api.BulkOperationResponse{
			Success:   false,
			Message:   "some categories failed",
			FailedIds: errStrings(errs),
		}, nil
	}

	return &api.BulkOperationResponse{
		Success: true,
		Message: "categories created successfully",
	}, nil
}

func (s *CategoryService) UpdateCategories(ctx context.Context, msg *api.UpdateCategoriesRequest) (*api.BulkOperationResponse, error) {
	var inputs []repository.UpdateCategoryInput
	for _, c := range msg.Categories {
		inputs = append(inputs, repository.UpdateCategoryInput{
			ID:   c.Id,
			Name: c.Name,
		})
	}

	errs := s.repo.Update(ctx, inputs)
	if len(errs) > 0 {
		return &api.BulkOperationResponse{
			Success:   false,
			Message:   "some updates failed",
			FailedIds: errStrings(errs),
		}, nil
	}

	return &api.BulkOperationResponse{
		Success: true,
		Message: "categories updated successfully",
	}, nil
}

func (s *CategoryService) DeleteCategories(ctx context.Context, msg *api.DeleteCategoriesRequest) (*api.BulkOperationResponse, error) {
	ids := msg.Ids
	if len(ids) == 0 {
		return nil, BadRequest("no ids provided")
	}

	errs := s.repo.Delete(ctx, ids)
	if len(errs) > 0 {
		return &api.BulkOperationResponse{
			Success:   false,
			Message:   "some deletions failed",
			FailedIds: errStrings(errs),
		}, nil
	}

	return &api.BulkOperationResponse{
		Success: true,
		Message: "categories deleted successfully",
	}, nil
}

File: src/service/category.rs
//! Categories service — business logic mirroring Go's `services/category.go`.

use std::sync::Arc;

use serde::Serialize;

use crate::error::{ApiError, Result};
use crate::repo::traits::category::{CategoryCreateInput, CategoryRow, CategoryUpdateInput};
use crate::repo::traits::CategoryRepo;
use crate::service::transaction::BulkResult;
use crate::timex::ts_rfc3339;

#[derive(Clone)]
pub struct CategoryService {
    repo: Arc<dyn CategoryRepo>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CategoryResponse {
    pub id: String,
    pub name: String,
    #[serde(rename = "createdAt")]
    pub created_at: String,
}

impl CategoryService {
    pub fn new(repo: Arc<dyn CategoryRepo>) -> Self {
        Self { repo }
    }

    pub async fn list(&self, user_id: &str, page_size: i64, page_token: &str) -> Result<(Vec<CategoryResponse>, String)> {
        let res = self.repo.list(user_id, page_size, page_token).await?;
        let items = res.categories.iter().map(category_wire).collect();
        Ok((items, res.next_page_token))
    }

    pub async fn create(&self, user_id: &str, names: &[String]) -> Result<BulkResult> {
        let inputs: Vec<_> = names.iter().filter(|n| !n.is_empty()).map(|n| CategoryCreateInput { name: n.clone() }).collect();
        match self.repo.create(user_id, &inputs).await {
            Ok(_) => Ok(BulkResult { success: true, message: "categories created successfully".to_string(), failed_ids: Vec::new(), skipped: 0 }),
            Err(e) => Ok(BulkResult { success: false, message: "some categories failed".to_string(), failed_ids: vec![e.message], skipped: 0 }),
        }
    }

    pub async fn update(&self, user_id: &str, inputs: &[(String, String)]) -> Result<BulkResult> {
        let repo_inputs: Vec<_> = inputs.iter().map(|(id, name)| CategoryUpdateInput { id: id.clone(), name: name.clone() }).collect();
        let errs = self.repo.update(user_id, &repo_inputs).await?;
        if !errs.is_empty() {
            return Ok(BulkResult { success: false, message: "some updates failed".to_string(), failed_ids: errs, skipped: 0 });
        }
        Ok(BulkResult { success: true, message: "categories updated successfully".to_string(), failed_ids: Vec::new(), skipped: 0 })
    }

    pub async fn delete(&self, user_id: &str, ids: &[String]) -> Result<BulkResult> {
        if ids.is_empty() {
            return Err(ApiError::bad_request("no ids provided"));
        }
        let errs = self.repo.delete(user_id, ids).await?;
        if !errs.is_empty() {
            return Ok(BulkResult { success: false, message: "some deletions failed".to_string(), failed_ids: errs, skipped: 0 });
        }
        Ok(BulkResult { success: true, message: "categories deleted successfully".to_string(), failed_ids: Vec::new(), skipped: 0 })
    }
}

fn category_wire(c: &CategoryRow) -> CategoryResponse {
    CategoryResponse { id: c.id.clone(), name: c.name.clone(), created_at: ts_rfc3339(&c.created_at) }
}
